features/lorawan/LoRaWANStack.cpp

```diff
@@ -0,0 +1,2046 @@
+    mib_set_request(&mib_req);
+
+    // Reset counters to zero. Will change in future with 1.1 support.
+    _lw_session.downlink_counter = 0;
+    _lw_session.uplink_counter = 0;
+
+    // Start loRaWAN state machine.
+    set_device_state(DEVICE_STATE_INIT);
+    return lora_state_machine();
+}
+
+/**
+ *
+ * Prepares the upload message to reserved ports
+ *
+ * \param port              Application port
+ * \return                  The length of data sent or a negative
+ *                          error code.
+ */
+int16_t LoRaWANStack::prepare_special_tx_frame(uint8_t port)
+{
+    switch (port) {
+    case 224:
+        if (_compliance_test.link_check == true) {
+            _compliance_test.link_check = false;
+            _compliance_test.state = 1;
+            _tx_msg.f_buffer_size = 3;
+            _tx_msg.f_buffer[0] = 5;
+            _tx_msg.f_buffer[1] = _compliance_test.demod_margin;
+            _tx_msg.f_buffer[2] = _compliance_test.nb_gateways;
+        } else {
+            switch (_compliance_test.state) {
+            case 4:
+                _compliance_test.state = 1;
+                break;
+            case 1:
+                _tx_msg.f_buffer_size = 2;
+                _tx_msg.f_buffer[0] = _compliance_test.downlink_counter >> 8;
+                _tx_msg.f_buffer[1] = _compliance_test.downlink_counter;
+                break;
+            }
+        }
+        break;
+    default:
+        break;
+    }
+
+    return _tx_msg.f_buffer_size;
+}
+
+uint16_t LoRaWANStack::check_possible_tx_size(uint16_t size)
+{
+    LoRaMacTxInfo_t txInfo;
+    if (LoRaMacQueryTxPossible(size, &txInfo) == LORAMAC_STATUS_LENGTH_ERROR) {
+        // Cannot transmit this much. Return how much data can be sent
+        // at the moment
+        return txInfo.MaxPossiblePayload;
+    }
+
+    return txInfo.CurrentPayloadSize;
+}
+
+/** Hands over the frame to MAC layer
+ *
+ * \return          returns the state of the LoRa MAC
+ */
+lora_mac_status_t LoRaWANStack::send_frame_to_mac()
+{
+    lora_mac_mcps_req_t mcps_req;
+    lora_mac_status_t status;
+   // LoRaMacTxInfo_t txInfo;
+
+    GetPhyParams_t phy_params;
+    PhyParam_t default_datarate;
+    phy_params.Attribute = PHY_DEF_TX_DR;
+    default_datarate = lora_phy.get_phy_params(&phy_params);
+
+    mcps_req.type = _tx_msg.type;
+
+    if (LORA_MCPS_UNCONFIRMED == mcps_req.type) {
+        mcps_req.req.unconfirmed.f_port = _tx_msg.message_u.unconfirmed.f_port;
+        mcps_req.f_buffer = _tx_msg.f_buffer;
+
+        mcps_req.f_buffer_size = _tx_msg.f_buffer_size;
+        mcps_req.req.unconfirmed.datarate = default_datarate.Value;
+
+    } else if (LORA_MCPS_CONFIRMED == mcps_req.type) {
+        mcps_req.req.confirmed.f_port = _tx_msg.message_u.confirmed.f_port;
+        mcps_req.f_buffer = _tx_msg.f_buffer;
+        mcps_req.f_buffer_size = _tx_msg.f_buffer_size;
+        mcps_req.req.confirmed.nb_trials =
+                _tx_msg.message_u.confirmed.nb_trials;
```

Comment:
Please make this statement in single line, makes it easier to read.

Reply:
Taken care off.